Firestore/core/src/firebase/firestore/util/equality.h

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_EQUALITY_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_EQUALITY_H_
+
+#include <memory>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+/**
+ * Checks if the values pointed to by two C++ pointers are equal. Two null
+ * pointers are also considered equal.
+ */
+template <typename T, typename = typename std::pointer_traits<T>::element_type>
```

Comment:
Interesting! I hadn't realized that `std::vector` would define `std::pointer_traits`, though now that you mention that, it makes sense since it's an array-like thing.

I wanted to disable this specifically in the case of Objective-C object references, which syntactically look like pointers but don't define `std::pointer_traits<FSTFoo*>::element_type`. The idea was that I might move the overload for Objective-C object references into this file (and namespace). The problem with removing this is that dereferencing an Objective-C object reference works. It looks like a pointer operation but is treated distinctly in Clang. Just basing this off the validity of `*T()` having a valid type doesn't work for that.

However, I didn't actually move `objc::Equals` into here so this can safely be deferred until (if) that becomes a reality. Maybe we'll be able to delete that instead.

In the meantime note that this code effectively implicitly asserts that `std::declval<T>() == nullptr` is valid because the code actually performs that operation. Replicating that in an explicit `static_assertion` seems like wasted effort. It would only be worthwhile to explicitly call this out to make it a specialization failure rather than an outright compilation error.

In the meantime, since there are no specializations amongst which to choose in this namespace I've merely removed the anonymous type.